torch/nn/parallel/distributed_c10d.py

```diff
@@ -0,0 +1,352 @@
+            bucket = self.buckets[bucket_idx][device_idx]
+            bucket.append(param.grad.data)
+
+            # We can flush these and save memory for replicas
+            if device_idx > 0:
+                param.grad = None
+                param.data.set_()
+
+            # Current device's bucket is full
+            if len(bucket) == self.bucket_sizes[bucket_idx]:
+                self.devs_ready[bucket_idx] += 1
+                if self.devs_ready[bucket_idx] < len(self.device_ids):
+                    return
+
+                if bucket_idx not in self.buckets_ready:
+                    self.buckets_ready.add(bucket_idx)
+                    if len(self.buckets_to_reduce) > 0 and self.buckets_to_reduce[0] == bucket_idx:
+                        # Reduce the bucket
+                        self._queue_reduction(bucket_idx)
+                        self.buckets_to_reduce.pop(0)
+
+                    # If all buckets are ready
+                    if len(self.buckets_ready) == len(self.bucket_order):
+                        # In case there are unreduced buckets, reduce them all
+                        if len(self.buckets_to_reduce) > 0:
+                            for b_idx in self.buckets_to_reduce:
+                                self._queue_reduction(b_idx)
+                        # A final sync for all the reduction works
+                        self._sync_reduction_works()
+
+        return distributed_data_parallel_hook
+
+    def _queue_reduction(self, bucket_idx):
+        grads_batch = self.buckets[bucket_idx]
+        grads_batch_coalesced = []
+
+        # coalesce the bucket
+        for dev_id, dev_grads_batch in zip(self.device_ids, grads_batch):
+            with torch.cuda.device(dev_id):
+                dev_grads_batch_coalesced = _flatten_dense_tensors(dev_grads_batch)
+                grads_batch_coalesced.append(dev_grads_batch_coalesced)
+
+        # reduce to the first GPU in self.device_ids
+        if len(self.device_ids) > 1:
+            nccl.reduce(grads_batch_coalesced, root=0, streams=self.default_streams)
+
+        # now work on the first gpu
+        reduction_work = c10d.all_reduce(grads_batch_coalesced[0], self.process_group)
+        self.reduction_works.append(reduction_work)
+        self.buckets_coalesced[bucket_idx] = grads_batch_coalesced[0]
+
+    def _sync_reduction_works(self):
+        for reduction_work in self.reduction_works:
+            reduction_work.wait()
+
+        # Now only work on the first GPU of self.device_ids, uncoalesce
+        # the gradients for each bucket
+        for bucket_idx, grads_batch in enumerate(self.buckets):
+            self.buckets_coalesced[bucket_idx] /= self.process_group.size()
+            grads_batch_reduced = _unflatten_dense_tensors(
+                self.buckets_coalesced[bucket_idx], grads_batch[0])
+
+            for grad, reduced in zip(grads_batch[0], grads_batch_reduced):
+                grad.copy_(reduced)
```

Comment:
So because we wait with the buffers until the very end, this basically doubles the memory usage, right? Also, it would be better to interleave calls to `wait()` with the copies (by putting it at the beginning of this loop), because if the next `wait()` would block you, then you might as well just continue queueing things that you know are ready.

Reply:
Yes, this basically double the memory usage.

Reply:
I interleaved in the second commit that I pushed